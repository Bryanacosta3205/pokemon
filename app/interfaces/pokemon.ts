export interface Pokemon {
  name: string;
  image: string;
  types: string[];
}

export interface Pagination {
  currentPage: number;
  totalPages: number;
  totalItems: number;
  itemsPerPage: number;
  hasNextPage: boolean;
  hasPreviousPage: boolean;
}

export interface PokeApiResponse {
  data: Pokemon[];
  pagination: Pagination;
  success: boolean;
}
