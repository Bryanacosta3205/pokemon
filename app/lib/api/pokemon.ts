import { fetcher } from "./client";

const API_BASE = process.env.NEXT_PUBLIC_BASE_URL;

export async function getPokemons(page = 1, limit = 50) {
  return fetcher(`${API_BASE}/pokemons?limit=${limit}&page=${page}`);
}
