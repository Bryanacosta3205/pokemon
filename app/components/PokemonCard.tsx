import { Card, Typography } from "@mui/material";
import { Pokemon } from "../interfaces/pokemon";
import PokemonTypeList from "./PokemonTypeList";
import Image from "next/image";

type Props = {
  pokemon: Pokemon;
};

const PokemonCard = ({ pokemon }: Props) => {
  return (
    <Card
      sx={{
        transition: "transform 0.2s, box-shadow 0.2s",
        "&:hover": {
          transform: "translateY(-4px)",
          boxShadow: 6,
        },
      }}
      style={{ backgroundColor: "#F2F2F2" }}
      className="h-80 flex flex-col px-3 py-2 border-6 border-neutral-200 "
    >
      <div className="flex justify-between flex-row px-0 mb-2">
        <Typography
          variant="h6"
          noWrap
          component="div"
          fontWeight={"bold"}
          align="center"
          sx={{
            textTransform: "capitalize",
            maxWidth: 120,
          }}
        >
          {pokemon.name}
        </Typography>
        <PokemonTypeList types={pokemon.types} />
      </div>
      <div
        className="flex justify-center"
        style={{
          background: `linear-gradient(
            135deg,
            rgba(255, 215, 0, 0.8) 0%,
            rgba(173, 216, 230, 0.8) 40%,
            rgba(255, 182, 193, 0.8) 100%
          )`,
          boxShadow: "0 4px 15px rgba(0, 0, 0, 0.1)",
        }}
      >
        <Image
          src={pokemon.image || "/placeholder.png"}
          alt={pokemon.name}
          height={120}
          width={140}
          preload={true}
        />
      </div>
    </Card>
  );
};

export default PokemonCard;
