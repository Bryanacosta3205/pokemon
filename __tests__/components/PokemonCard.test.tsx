import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import PokemonCard from "@/app/components/PokemonCard";
import { Pokemon } from "@/app/interfaces/pokemon";

jest.mock("@/app/components/PokemonTypeList", () => {
  return function MockPokemonTypeList({ types }: any) {
    return (
      <div data-testid="pokemon-type-list">
        {types.map((type: string) => (
          <span key={type} data-testid={`type-${type}`}>
            {type}
          </span>
        ))}
      </div>
    );
  };
});

describe("PokemonCard", () => {
  const mockPokemon: Pokemon = {
    name: "bulbasaur",
    image: "https://example.com/bulbasaur.png",
    types: ["grass", "poison"],
  };

  describe("Rendering", () => {
    it("renders the pokemon card", () => {
      render(<PokemonCard pokemon={mockPokemon} />);

      const card = screen.getByRole("img").closest(".MuiCard-root");
      expect(card).toBeInTheDocument();
    });

    it("displays the pokemon name", () => {
      render(<PokemonCard pokemon={mockPokemon} />);

      expect(screen.getByText("bulbasaur")).toBeInTheDocument();
    });

    it("capitalizes the pokemon name with CSS", () => {
      render(<PokemonCard pokemon={mockPokemon} />);

      const nameElement = screen.getByText("bulbasaur");
      expect(nameElement).toHaveStyle({ textTransform: "capitalize" });
    });

    it("displays the pokemon image", () => {
      render(<PokemonCard pokemon={mockPokemon} />);

      const image = screen.getByRole("img", { name: /bulbasaur/i });
      expect(image).toBeInTheDocument();
      expect(image).toHaveAttribute("src", mockPokemon.image);
      expect(image).toHaveAttribute("alt", mockPokemon.name);
    });

    it("renders PokemonTypeList with correct types", () => {
      render(<PokemonCard pokemon={mockPokemon} />);

      expect(screen.getByTestId("pokemon-type-list")).toBeInTheDocument();
      expect(screen.getByTestId("type-grass")).toBeInTheDocument();
      expect(screen.getByTestId("type-poison")).toBeInTheDocument();
    });
  });

  describe("Different Pokemon Data", () => {
    it("renders pokemon with single type", () => {
      const singleTypePokemon: Pokemon = {
        name: "charmander",
        image: "https://example.com/charmander.png",
        types: ["fire"],
      };

      render(<PokemonCard pokemon={singleTypePokemon} />);

      expect(screen.getByText("charmander")).toBeInTheDocument();
      expect(screen.getByTestId("type-fire")).toBeInTheDocument();
    });

    it("renders pokemon with long name", () => {
      const longNamePokemon: Pokemon = {
        name: "fletchinder",
        image: "https://example.com/fletchinder.png",
        types: ["fire", "flying"],
      };

      render(<PokemonCard pokemon={longNamePokemon} />);

      expect(screen.getByText("fletchinder")).toBeInTheDocument();
    });

    it("renders pokemon with multiple types", () => {
      const multiTypePokemon: Pokemon = {
        name: "charizard",
        image: "https://example.com/charizard.png",
        types: ["fire", "flying"],
      };

      render(<PokemonCard pokemon={multiTypePokemon} />);

      expect(screen.getByTestId("type-fire")).toBeInTheDocument();
      expect(screen.getByTestId("type-flying")).toBeInTheDocument();
    });

    it("handles pokemon with hyphenated name", () => {
      const hyphenatedPokemon: Pokemon = {
        name: "farfetch-d",
        image: "https://example.com/farfetchd.png",
        types: ["normal", "flying"],
      };

      render(<PokemonCard pokemon={hyphenatedPokemon} />);

      expect(screen.getByText("farfetch-d")).toBeInTheDocument();
    });
  });

  describe("Props Validation", () => {
    it("passes correct types to PokemonTypeList", () => {
      render(<PokemonCard pokemon={mockPokemon} />);

      expect(screen.getByTestId("type-grass")).toBeInTheDocument();
      expect(screen.getByTestId("type-poison")).toBeInTheDocument();
    });

    it("uses pokemon name for image alt text", () => {
      render(<PokemonCard pokemon={mockPokemon} />);

      const image = screen.getByAltText("bulbasaur");
      expect(image).toBeInTheDocument();
    });

    it("uses pokemon image URL correctly", () => {
      render(<PokemonCard pokemon={mockPokemon} />);

      const image = screen.getByRole("img");
      expect(image).toHaveAttribute("src", "https://example.com/bulbasaur.png");
    });
  });

  describe("Accessibility", () => {
    it("provides accessible image alt text", () => {
      render(<PokemonCard pokemon={mockPokemon} />);

      const image = screen.getByRole("img", { name: /bulbasaur/i });
      expect(image).toBeInTheDocument();
    });
  });

  describe("Snapshot Tests", () => {
    it("matches snapshot for basic pokemon", () => {
      const { container } = render(<PokemonCard pokemon={mockPokemon} />);
      expect(container).toMatchSnapshot();
    });

    it("matches snapshot for single-type pokemon", () => {
      const singleTypePokemon: Pokemon = {
        name: "pikachu",
        image: "https://example.com/pikachu.png",
        types: ["electric"],
      };

      const { container } = render(<PokemonCard pokemon={singleTypePokemon} />);
      expect(container).toMatchSnapshot();
    });

    it("matches snapshot for multi-type pokemon", () => {
      const multiTypePokemon: Pokemon = {
        name: "charizard",
        image: "https://example.com/charizard.png",
        types: ["fire", "flying"],
      };

      const { container } = render(<PokemonCard pokemon={multiTypePokemon} />);
      expect(container).toMatchSnapshot();
    });
  });
});
