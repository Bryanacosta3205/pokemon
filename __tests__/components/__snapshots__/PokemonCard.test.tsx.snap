// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`PokemonCard Snapshot Tests matches snapshot for basic pokemon 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root h-80 flex flex-col px-3 py-2 border-6 border-neutral-200 css-i2mtdz-MuiPaper-root-MuiCard-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12); background-color: rgb(242, 242, 242);"
  >
    <div
      class="flex justify-between flex-row px-0 mb-2"
    >
      <div
        class="MuiTypography-root MuiTypography-h6 MuiTypography-alignCenter css-rc5xui-MuiTypography-root"
        style="--Typography-textAlign: center;"
      >
        bulbasaur
      </div>
      <div
        data-testid="pokemon-type-list"
      >
        <span
          data-testid="type-grass"
        >
          grass
        </span>
        <span
          data-testid="type-poison"
        >
          poison
        </span>
      </div>
    </div>
    <img
      alt="bulbasaur"
      class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-1j9c31z-MuiCardMedia-root"
      src="https://example.com/bulbasaur.png"
      style="background: linear-gradient( 135deg, rgba(255, 215, 0, 0.8) 0%, rgba(173, 216, 230, 0.8) 40%, rgba(255, 182, 193, 0.8) 100% ); box-shadow: 0 4px 15px rgba(0, 0, 0, 0.1);"
    />
  </div>
</div>
`;

exports[`PokemonCard Snapshot Tests matches snapshot for multi-type pokemon 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root h-80 flex flex-col px-3 py-2 border-6 border-neutral-200 css-i2mtdz-MuiPaper-root-MuiCard-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12); background-color: rgb(242, 242, 242);"
  >
    <div
      class="flex justify-between flex-row px-0 mb-2"
    >
      <div
        class="MuiTypography-root MuiTypography-h6 MuiTypography-alignCenter css-rc5xui-MuiTypography-root"
        style="--Typography-textAlign: center;"
      >
        charizard
      </div>
      <div
        data-testid="pokemon-type-list"
      >
        <span
          data-testid="type-fire"
        >
          fire
        </span>
        <span
          data-testid="type-flying"
        >
          flying
        </span>
      </div>
    </div>
    <img
      alt="charizard"
      class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-1j9c31z-MuiCardMedia-root"
      src="https://example.com/charizard.png"
      style="background: linear-gradient( 135deg, rgba(255, 215, 0, 0.8) 0%, rgba(173, 216, 230, 0.8) 40%, rgba(255, 182, 193, 0.8) 100% ); box-shadow: 0 4px 15px rgba(0, 0, 0, 0.1);"
    />
  </div>
</div>
`;

exports[`PokemonCard Snapshot Tests matches snapshot for single-type pokemon 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root h-80 flex flex-col px-3 py-2 border-6 border-neutral-200 css-i2mtdz-MuiPaper-root-MuiCard-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12); background-color: rgb(242, 242, 242);"
  >
    <div
      class="flex justify-between flex-row px-0 mb-2"
    >
      <div
        class="MuiTypography-root MuiTypography-h6 MuiTypography-alignCenter css-rc5xui-MuiTypography-root"
        style="--Typography-textAlign: center;"
      >
        pikachu
      </div>
      <div
        data-testid="pokemon-type-list"
      >
        <span
          data-testid="type-electric"
        >
          electric
        </span>
      </div>
    </div>
    <img
      alt="pikachu"
      class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-1j9c31z-MuiCardMedia-root"
      src="https://example.com/pikachu.png"
      style="background: linear-gradient( 135deg, rgba(255, 215, 0, 0.8) 0%, rgba(173, 216, 230, 0.8) 40%, rgba(255, 182, 193, 0.8) 100% ); box-shadow: 0 4px 15px rgba(0, 0, 0, 0.1);"
    />
  </div>
</div>
`;
